### DISCUSSION TITLE
Node is either not visible or not an HTMLElement

### DISCUSSION THREAD

**Discussion by Mephisto090** [cefsharp/CefSharp.Dom#17] [Q&A] [answered]
Hello dear developer,

I get the error message: CefSharp.Puppeteer.PuppeteerException: "Node is either not visible or not an HTMLElement"

What does this error mean?

**Comment by Mephisto090** [comment, marked as answer]
I found the error myself. If an element is not visible and hidden by Javascript, it will not find it. Should that be so?

**Reply by amaitland** [reply]
> If an element is not visible and hidden by Javascript, it will not find it. Should that be so?

Yes. This is expected. ClickAsync simulates a user click, a user cannot click a hidden element. 

I've already posted an alternative in the example I provided. https://github.com/amaitland/Mephisto090.MinimalChrome/blob/master/MinimalChrome/Form1.vb#L60

**Reply by Mephisto090** [reply]
I have a nice idea for you. Some websites read where the button was pressed. I think the button is now pressed at position 0:0 which would not be human. do you think it would be a good idea to somehow make it possible that it would be more human?

**Reply by amaitland** [reply]
> I think the button is now pressed at position 0:0 

What are you basing this off? What debugging did you do to come to this conclusion. 

The code should already click in the middle of the clickable area. The click should occur in the middle of the clickable area. 

https://github.com/cefsharp/Puppeteer/blob/main/lib/PuppeteerSharp/ElementHandle.cs#L699

**Reply by amaitland** [reply]
Puppeteer supports specifying a custom offset that's something that might be worth supporting. 

https://github.com/puppeteer/puppeteer/blob/main/src/common/JSHandle.ts#L536

Pull requests welcome

**Reply by Mephisto090** [reply]
> What are you basing this off? What debugging did you do to come to this conclusion.

That was just an assumption on my part. I mean, with a 50px x 50px button, the middle is of course also very noticeable.


On a page that I know of, two inputs are filled by Javascript. Once the X and once the Y coordinate. I can send you the source code as an excerpt (the 2 inputs of X and Y)

Either you make it so that the user can choose where to press or Puppeteer finds out from where to where to press and makes it random.

**Reply by amaitland** [reply]
> the middle is of course also very noticeable.

What problem are you actually seeing from the current behaviour? 

> On a page that I know of, two inputs are filled by Javascript. Once the X and once the Y coordinate. I can send you the source code as an excerpt (the 2 inputs of X and Y)

I have absolutely no idea what you mean by this. 

> Either you make it so that the user can choose where to press or Puppeteer finds out from where to where to press and makes it random.

Matching the puppeteer behaviour is likely the path we'll take.

If you require this functionality you are welcome to implement and submit a pull request. Porting the unit tests will be required.

**Reply by Mephisto090** [reply]
> What problem are you actually seeing from the current behaviour?

The site I know records which buttons were pressed, how and where. I think this is a protection against bots. Always exactly in the same place is just not human and so a bot would be noticed

> I have absolutely no idea what you mean by this.

I'll just show you source code for it. i think that explains it (I've removed unnecessary code from this example)

```
<a id="ButtonID" data-tooltip-theme="0" rel="nofollow"><span class="text new-console">Buttonname</span>
<form class="classname" action="javascript:;" method="post" id="formbutton">
<input name="input_x" id="input_x" type="hidden" value="0">
<input name="input_y" id="input_y" type="hidden" value="0">
</form></a>
```

If we now have a button that is 50px x 50px and we press it at X: 20 and Y: 32 then the input fields are filled via Javascript as in the example below

```
<input name="input_x" id="input_x" type="hidden" value="20">
<input name="input_y" id="input_y" type="hidden" value="32">
```

Thus, the page knows how the click behavior is

> If you require this functionality you are welcome to implement and submit a pull request. Porting the unit tests will be required.

How can I do that? I'm sorry that I'm not that familiar with guthub. Never used it before you :(
If possible please use plain English :)

**Reply by amaitland** [reply]
You can calculate your own set of coordinates and click the mouse

https://github.com/cefsharp/Puppeteer/blob/main/lib/PuppeteerSharp.Tests/MouseTests/MouseTests.cs#L50

**Reply by Mephisto090** [reply]
That you're reprogramming it now? I'll try it later

**Reply by Mephisto090** [reply]
```
Dim Element As ElementHandle() = Await DevtoolsContext.QuerySelectorAllAsync(TagName)
For Each Elements In Element
  Await Elements.ClickAsync(25, 30)
  Await DevtoolsContext.Mouse.ClickAsync(25, 30)
Next
```


Does this also work in the `For Each` variant?
The `ElementHandle` does not know `Mouse.ClickAsync`

**Reply by amaitland** [reply]
> That you're reprogramming it now? 

I will port the offset feature from puppeteer for $200USD payable via GitHub sponsors if you would like. 

You are welcome to implement the feature yourself and submit a pull request if that's your preference. 

> Does this also work in the For Each variant?
The ElementHandle does not know Mouse.ClickAsync

This hasn't been implemented yet.

**Reply by Mephisto090** [reply]
> You can calculate your own set of coordinates and click the mouse
> 
> https://github.com/cefsharp/Puppeteer/blob/main/lib/PuppeteerSharp.Tests/MouseTests/MouseTests.cs#L50

```
Dim Element As ElementHandle() = Await DevtoolsContext.QuerySelectorAllAsync(TagName)
For Each Elements In Element
  Await Elements.ClickAsync(25, 30)
Next
```

Where is the `Elements.ClickAsync()` function?
I'll try to incorporate it so that it's not always pressed in the middle

**Reply by Mephisto090** [reply]
I would like to add another feature where the user can choose the button click position.

In the `ClickablePointAsync` function.
is there `x` and `y` the maximum size of the button?
